db/migrations/00003_add_edges_functions.sql

```diff
@@ -0,0 +1,418 @@
+    message           JSONB;     -- Parsed Message JSON
+    item              JSONB;     -- Individual item in JSON arrays
+BEGIN
+    -- Extract the Message field as JSON
+    message := (label_variant_value::jsonb ->> 'Message')::jsonb;
+
+    -- Add edge to provider
+    IF (message ->> 'provider') IS NOT NULL THEN
+        PERFORM add_edge(
+            provider_type,
+            message ->> 'provider',
+            '',
+            label_type,
+            parse_resource_key(label_variant_key, 'name'),
+            parse_resource_key(label_variant_key, 'variant')
+        );
+    END IF;
+
+    -- Process trainingsets array
+    IF message -> 'trainingsets' IS NOT NULL THEN
+        FOR item IN SELECT jsonb_array_elements(message -> 'trainingsets')
+        LOOP
+            PERFORM add_edge(
+                label_type,
+                parse_resource_key(label_variant_key, 'name'),
+                parse_resource_key(label_variant_key, 'variant'),
+                training_set_type,
+                item ->> 'name',
+                item ->> 'variant'
+            );
+        END LOOP;
+    END IF;
+
+    -- Add edge to source
+    IF message -> 'source' IS NOT NULL THEN
+        PERFORM add_edge(
+            source_type,
+            ((message -> 'source')::jsonb) ->> 'name',
+            ((message -> 'source')::jsonb) ->> 'variant',
+            label_type,
+            parse_resource_key(label_variant_key, 'name'),
+            parse_resource_key(label_variant_key, 'variant')
+        );
+    END IF;
+END;
+$$;
+
+-- Function: process_source_variant
+CREATE FUNCTION process_source_variant(
+    source_variant_key text,
+    source_variant_value text
+) 
+RETURNS void
+LANGUAGE plpgsql
+AS $$
+DECLARE
+    source_type       INT := 7;  -- Resource type for source_variant
+    provider_type     INT := 8;  -- Resource type for provider
+    feature_type      INT := 4;  -- Resource type for feature_variant
+    training_set_type INT := 6;  -- Resource type for training_set_variant
+    label_type        INT := 5;  -- Resource type for label_variant
+    message           JSONB;     -- Parsed Message JSON
+    item              JSONB;     -- Individual item in lists
+BEGIN
+    -- Extract the Message field as JSON
+    message := (source_variant_value::jsonb ->> 'Message')::jsonb;
+
+    -- Add an edge from SOURCE_VARIANT to its provider
+    IF (message ->> 'provider') IS NOT NULL THEN
+        PERFORM add_edge(
+            provider_type,
+            message ->> 'provider',
+            '',
+            source_type,
+            parse_resource_key(source_variant_key, 'name'),
+            parse_resource_key(source_variant_key, 'variant')
+        );
+    END IF;
+
+    -- Process the "source" array in SQLTransformation or DFTransformation
+    IF message -> 'transformation' -> 'SQLTransformation' -> 'source' IS NOT NULL THEN
+        FOR item IN SELECT jsonb_array_elements(message -> 'transformation' -> 'SQLTransformation' -> 'source')
```

Comment:
can't believe this isn't also called inputs.
Are we sure these are set before they are passing into psql?